### DISCUSSION TITLE
Budgets and Reports with multiple currencies

### DISCUSSION THREAD

**Discussion by evanre** [firefly-iii/firefly-iii#5639] [Firefly III]
Hey, I've recently relocated to another country with different currency. I've changed my default currency from UAH to PLN. I've also opened a couple of accounts in PLN. But I still have many expenses in my origin's country currency (UAH). Furthermore, I'm using my old CC's (UAH) in a new country with PLN. Regarding filling transactions, everything's working just great, thanks to "Foreign currency amount" field.

But I came to with Reports and Budgets thing, that now everything is being displayed in two currencies. 
- I can't check the stats of my previous expenses, almost every category has two records (for UAH and PLN), even though I've precisely indicated in each "foreign" expense both amounts (UAH and PLN).
- I'm also not sure how to use budgets. For example, I'm going to spend 2000pln/month on food, but the expanses can be either from PLN or UAH accounts (for UAH, I also provide foreign amount).

Any advices are greatly appreciated :)

**Comment by Trevasotilo** [comment]
Yes I would also like to know about this, many personal finance programs break down at exactly this problem

**Comment by JC5** [comment]
State of the art, I'm afraid. 

> But I came to with Reports and Budgets thing, that now everything is being displayed in two currencies.

There is currently no option to (automatically) convert from one currency to the other. Everything remains split.

> For example, I'm going to spend 2000pln/month on food, but the expanses can be either from PLN or UAH accounts (for UAH, I also provide foreign amount).

You can set up budgets for multiple currencies.

**Comment by evanre** [comment]
@JC5 Thanks for answer. So, what currently "Foreign currency" field does? I see that in budgets, in "already spent" section some transactions displays in both currencies. I mean, if budget "Food" has only one multicurrency transaction for 100pln/700uah, it shows up both values in "Already spend".

**Reply by JC5** [reply]
I can't choose for you which budget it should apply to. I am getting mixed requests from users.

**Comment by dklymenk** [comment]
I actually have a similar (or even the same) problem. Even though I have been living with my spouse in a foreign country for a while now, my wife still pays for some of the stuff with her old card that is in a different currency.

I think I can rephrase the OPs questions, so it's more clear what's the expected result such that it won't imply any guess work from the project's maintainer.

They way I understand it, the budget view already uses the foreign amount field where possible for the "Spent in budget" calculation. So to answer's OPs 2nd question about budgets. If you specify let's say your grocery budget to be in PLN, and every transaction towards that budget is going be either in PLN or have PLN value in a foreign amount field, the calculation and the chart on the dashboard will use the PLN values for every transaction.

This doesn't seem to be the case for category bar chart on dashboard and category pie charts on other pages. The foreign amount field is not used there at all. The way I personally see it, there should be no issue to use the default currency if a transaction has one similarly to budgets. This is only way I can see those charts being useful in a multi-currency environment without doing any actual conversions.

I will try to explain expected result using the demo site.

The default grocery budget there is 100 EUR and there is already 48.90 EUR spent from it. Additionally there is a category "daily groceries" that has 42.12 EUR worth of transactions on it.

![DeepinScreenshot_select-area_20220202235724](https://user-images.githubusercontent.com/64093836/152236210-1eaf5fcb-e3b5-4d8b-877a-21116ad8e255.png)
![DeepinScreenshot_select-area_20220202235731](https://user-images.githubusercontent.com/64093836/152236206-36d9a53e-235f-423b-8a61-639997a2fafb.png)

Now let's say you go to a go grocery store and your EUR card declines, or you forgot it, or you moved too much money from it to your savings account, and you resort to paying with your USD card. A transaction would look like this:
![DeepinScreenshot_select-area_20220203000218](https://user-images.githubusercontent.com/64093836/152237023-0bee971c-62b9-4ea2-baa2-6b4428436e4d.png)

You can see on the dashboard that the budget calculation used the EUR value and now shows 48.90 + 10 = 58.90 EUR.
![DeepinScreenshot_select-area_20220203000355](https://user-images.githubusercontent.com/64093836/152237177-d2f63803-7f42-4920-aabf-40408e6230ee.png)

However, the category bar chart now shows a USD bar for the daily groceries category and the EUR bar doesn't seem to be affected by the foreign amount whatsoever.
![DeepinScreenshot_select-area_20220203000401](https://user-images.githubusercontent.com/64093836/152237555-e3bbaa4d-f8b9-4668-bbc1-5ee8bffceb54.png)

I don't think any case can be made for not using a default currency wherever possible. There has been a number of issues reported about these charts being useless for people dealing in multiple currencies. It is especially the case when dealing in currencies that have a big conversion rate (making the stronger currency bar barely visible in most of the cases). So this change would make the charts better for people using accounts in different currencies, while it won't affect the single currency people at all. Are there any downsides I missed?

**Reply by Trevasotilo** [reply]
I am not a developer, but I can't imagine that it would be very hard to implement something like an option to combine transactions from one category with multiple currencies into a report of only one currency according to user's defined exchange ratio. This is especially useful when looking at reports. (no need to pull exchange data from the web or anything like that)

In some other money managers I even saw that on the report it adds them on top of eachother (ex. 100$ food purchase, and 10€ food in reports it shows up as 110$ expenditure), and in some other, they completely ignore non default currency expenditures.

**Reply by JC5** [reply]
Right now, the strategy is to split multiple currencies as much as possible. Double bars, double charts, double report entries, everything. Any time you see money being mixed up, that's the bug. Several charts suffer from this issue (I'm aware).

Only when Firefly III can actually properly distinguish between currencies, can you think about merging them and setting an exchange rate.

The idea of letting a user set the exchange rate has been suggested and is on my list.

**Reply by dklymenk** [reply]
Are you trying to say that foreign amount field in the context of withdrawals is meant to be entirely cosmetic? Meaning it shouldn't impact any charts or calculations. I think we're talking about different things. Please let me clear it up.

> Right now, the strategy is to split multiple currencies as much as possible. Double bars, double charts, double report entries, everything. Any time you see money being mixed up, that's the bug. Several charts suffer from this issue (I'm aware).

I'm pretty sure it's not a bug, the behaviour with budgets I described above was requested as a feature and was implemented as a feature (https://github.com/firefly-iii/firefly-iii/issues/3875).

> Only when Firefly III can actually properly distinguish between currencies, can you think about merging them and setting an exchange rate.

The behaviour I describe above doesn't need to rely on any exchange rates. If your transactions already have foreign amount in your default currency that is better and more accurate than any global exchange rate. The fact that this behaviour was implemented as a feature for budgets, but not for other charts seems like an inconsistency to me, but definitely not a bug. What do you think?

**Reply by JC5** [reply]
Let me stress that the foreign amount is not cosmetic.

> The fact that this behaviour was implemented as a feature for budgets, but not for other charts seems like an inconsistency to me, but definitely not a bug. What do you think?

This is indeed an inconsistency I plan to fix.

**Reply by dklymenk** [reply]
Great! Thank you for your work.

**Comment by evanre** [comment]
Here's my point. We have a "Groceries" budget, with 3 transactions in it. One of the transactions was done in EUR but paid by GBP asset. The problem that budget sums both values as separate transactions, and thus I can't forecast multiple-currencies budget. From my cards **was spent 14GBP and 10EUR**, budget **shows I spent 14GBP and 20EUR**.

If calculating exchange rate from "Amount/Foreign Amount" is not an option, then we should not sum both values from multicurrency transaction as separate spending.
![CleanShot 2022-02-03 at 10 09 56](https://user-images.githubusercontent.com/12615906/152314240-94f1e8b4-1b46-4418-956c-a17172af679e.png)
![CleanShot 2022-02-03 at 10 11 13](https://user-images.githubusercontent.com/12615906/152314254-8cb1612c-9c34-47f1-a594-b6f6707c17ed.png)

**Reply by JC5** [reply]
But I see two transactions in your screenshot? That sums up to 20.

**Reply by dklymenk** [reply]
@JC5 I think the problem he has is that the transaction where he specified the foreign amount effectively counted towards both EUR and GBP spending.

**Reply by JC5** [reply]
Yes I know. What did you expect? Should I just ignore one of these two amounts?

**Reply by dklymenk** [reply]
@evanre There are 2 workarounds I can suggest you:
1. Make the budget multi-currency and never specify the foreign amount to prevent duplication. I personally don't like this one, because it is highly inconvenient to make sense of a state of a budget like that when you go over it in one currency, but have some unspent amount in another currency.
![DeepinScreenshot_select-area_20220203124118](https://user-images.githubusercontent.com/64093836/152317833-d9fec9b4-0199-40b8-8159-e2087c973556.png)
2. Always set the foreign amount in your primary currency for withdrawals that use asset accounts not in your primary currency. This way you can set the budget in your primary currency and pay attention only to math that uses that currency.

**Reply by dklymenk** [reply]
> Yes I know. What did you expect? Should I just ignore one of these two amounts?

I guess one would expect the transaction that has a foreign currency specified to be only included in a calculation for primary budget currency, but I can see why it would produce a lot of confusion for other cases like:
* having a budget in currency A, but a transaction in currency B and C (foreign amount). Maybe you only use currency B here
* having a budget in currencies A and B, and a transaction in currency A and B (foreign amount). Which currency is used here? Your primary profile currency? What if your primary profile currency is neither A nor B?

Honestly trying to find a universal consistent logic that would solve this particular issue sounds like a major headache. I think the solution 2 in my post above is simple and universal enough. Every time you buy something in foreign currency you usually do the math of how much it is in your primary currency anyway.

**Reply by evanre** [reply]
@JC5 If you have multiple cards in different currencies (EUR, GBP) and pay in different ways (GBP»GBP, GBP»EUR, EUR»EUR, EUR»GBP), it's not possible to calculate the budgets because everything is summing up. On screenshot, I showed that "really" I spent 14GBP and 10EUR, but budget counted that I spent 14GBP and 20EUR.

Let's say the GBP is my primary currency, then if I specify additional budget in EUR, it should count only "EUR»GBP" spending, and not "GBP>EUR", since it's already counted in GBP budget.

**Reply by dklymenk** [reply]
> Let's say the GBP is my primary currency, then if I specify additional budget in EUR, it should count only "EUR»GBP" spending, and not "GBP>EUR", since it's already counted in GBP budget.

This makes sense, but it's still hard to come up with an intuitive solution for this problem that would work for all users. For this particular case if you don't want to use a single currency budget (my solution 2 from above), I'd say you can create 2 budgets (one in EUR and GBP) instead of one multi-currency one. You will get the exact same functionality from it, but it will allow you to only pay attention to math in primary budget currency.

**Reply by JC5** [reply]
Yes, the challenge is that this use case has to work for everybody, so I'm reluctant to make the choice as you suggest it. I had other users argue the case the other way around.

**Reply by evanre** [reply]
I understand. Thanks for your time, guys. I end up creating a Google Spreadshit for budget planning that fits my needs. But it doubles effort on planning, so I think I'll go with [solution #2](https://github.com/firefly-iii/firefly-iii/discussions/5639#discussioncomment-2102475) from @dklymenk.
